商品購入機能

jujuko75 opened the pull request:
### What
最終課題フリマアプリのタスク「商品購入機能」を実装しましたので、レビューをお願いします。
商品詳細ページから商品購入ページへ遷移し、クレカ情報及び商品送付先情報を入力し購入を完了するもの。

### Why
商品決済及び配送先情報の保存のため


**以下、実装条件及び稼働確認エビデンスです**

【前提条件】
◆未購入商品情報
・出品者：テスト太郎
・購入ページ： http://localhost:3000/items/11/order/buyerinfos

◆購入済商品情報
・出品者：テスト次郎
・購入ページ： http://localhost:3000/items/13/order/buyerinfos
　

【ログイン前】
・ログアウトの場合、商品購入ページURLを直接入力するとログインページに遷移する
　未購入商品の場合：
　　[https://gyazo.com/644ca8f08bcdc820ac48a6fd9b4c0d16](https://gyazo.com/644ca8f08bcdc820ac48a6fd9b4c0d16)
　
　購入済商品の場合：
　　[https://gyazo.com/f301d2eaf59b57a747b2e788b6b5d0b6](https://gyazo.com/f301d2eaf59b57a747b2e788b6b5d0b6)
	

【ログインユーザ = テスト太郎】
・ログインユーザが自身の出品した商品の購入ページURLを直接入力するとトップページに遷移する
　①自身の出品した商品の場合、商品購入ページのリンクが表示されない
　　[https://gyazo.com/23ac53f263f7fecc71e07e24ac27e5d5](https://gyazo.com/23ac53f263f7fecc71e07e24ac27e5d5)　
　②直接URLを入力した場合トップページへ遷移し、商品購入ページは表示されない。　　　
　　[https://gyazo.com/e3086c404424c668cc3992dedce806f6](https://gyazo.com/e3086c404424c668cc3992dedce806f6)

・ログインユーザが既に売却済の商品の商品購入ページURLを直接入力するとトップページに遷移する
　①売却済の商品画面
　　[https://gyazo.com/900ec38a9ffef7a19b862fc9b3ab9958](https://gyazo.com/900ec38a9ffef7a19b862fc9b3ab9958)
　②直接URLを入力した場合トップページへ遷移し、商品購入ページは表示されない。
　　[https://gyazo.com/fe96ba95ae8c60a0d2e1dc536479dcbf](https://gyazo.com/fe96ba95ae8c60a0d2e1dc536479dcbf)


【ログインユーザ = テスト次郎】
・ログインユーザが自身が出品した商品であっても売却済の場合は「編集・削除」が表示されない。
　　[https://gyazo.com/0a9b83b167314bbeab0f7e6b1c1f2203](https://gyazo.com/0a9b83b167314bbeab0f7e6b1c1f2203)

・商品購入ページで入力に不備がある場合、エラーメッセージが出力され、購入ページへ戻ること。
　　①購入ページへ遷移する
　　　[https://gyazo.com/29b33d0bca23c5611ec7b38941f3815a](https://gyazo.com/29b33d0bca23c5611ec7b38941f3815a)
　　②入力不備状態で購入ボタンを押下する。
　　　[https://gyazo.com/bf20d70d3727828084fddb15a12bce5f](https://gyazo.com/bf20d70d3727828084fddb15a12bce5f)

**◎（正常動作確認）商品購入**
　①購入する商品を選択し、商品購入ページへ遷移する。
　　　[https://gyazo.com/2e08376a8ea87ad1d5048c747a2a9f3d](https://gyazo.com/2e08376a8ea87ad1d5048c747a2a9f3d)
　②クレカ情報、商品送付先情報を入力し、購入ボタンを押下する。トップ画面へ遷移する。
　　　[https://gyazo.com/6b412a73805ab9dd0d1af777b2fd0e80](https://gyazo.com/6b412a73805ab9dd0d1af777b2fd0e80)
　③②で購入した商品表示（SoldOut）
　　　[https://gyazo.com/d9b7f5e32186f60994713e97155945de](https://gyazo.com/d9b7f5e32186f60994713e97155945de)
　④pay.jp売上
　　　[https://gyazo.com/3a07b9b3db1727db6b9154833c203d86](https://gyazo.com/3a07b9b3db1727db6b9154833c203d86)


【単体テスト結果】
[https://gyazo.com/f237bae65884b5b226676634b1fbd5e3](https://gyazo.com/f237bae65884b5b226676634b1fbd5e3)

jujuko75 committed:
購入者情報入力機能

jujuko75 committed:
クレジット決済機能追加

※中間テーブル作成（purchase→orderへ変更）

jujuko75 committed:
購入履歴管理テーブル作成

①購入履歴管理テーブル　:order
②フォームオブジェクト（order_buyerinfo）の作成
 →order,buyerinfoへの保存

jujuko75 committed:
購入済（Sold out）表示の条件分岐機能追加

jujuko75 committed:
購入済（Sold out）表示の条件分岐機能追加2

jujuko75 committed:
単体テスト、条件分岐追加

条件分岐（ログイン有無、売却済商品購入ページへの遷移）追加

kuu0123 commented on app/controllers/buyerinfos_controller.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  def create
+    set_item
```
```ruby
	    set_item
```
上記はメソッドを呼び出すのではなく、before_actionにset_item追加することで適用しましょう！

jujuko75 replied:
コメントありがとうございます。ご指摘の内容で修正しました。

kuu0123 commented on app/controllers/buyerinfos_controller.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def index
+    @order_buyerinfo = OrderBuyerinfo.new
+    set_item
```
```ruby
	    set_item
```
上記はメソッドを呼び出すのではなく、before_actionにset_item追加することで適用しましょう！

jujuko75 replied:
コメントありがとうございます。ご指摘の内容で修正しました。

kuu0123 commented on app/views/items/show.html.erb:
```diff
@@ -25,13 +25,15 @@
+        <% if !@item.order.present?%>
     <%= link_to '商品の編集', edit_item_path(@item.id), method: :get, class: "item-red-btn" %>
     <p class='or-text'>or</p>
     <%= link_to '削除', item_path(@item.id), method: :delete, class:'item-destroy' %>
+        <% end%>
       <% else %>
-    <%# 商品が売れていない場合はこちらを表示しましょう → 商品購入機能実装後 %>
-    <%= link_to '購入画面に進む', "#" ,class:"item-red-btn"%>
-    <%# //商品が売れていない場合はこちらを表示しましょう → 商品購入機能実装後 %>
+        <% if !@item.order.present?%>
```
```ruby
<% if !@item.order.present?%>
```
実装方針としては正解ですが、28行目と34行目の記述が重複しているようなので、こちらはifとelseどちらにも適用されるように、26行目に条件を追加してはいかがでしょうか？

jujuko75 replied:
コメントいただいた内容で修正しました。

kuu0123 commented on spec/models/order_buyerinfo_spec.rb:
```diff
@@ -0,0 +1,69 @@
+        @order_buyerinfo.valid?
+        expect(@order_buyerinfo.errors.full_messages).to include("House number can't be blank")
+      end
+      it '電話番号が空だと保存できないこと' do
```
電話番号について、以下のテストも実装しましょう！
```
・英数混合では登録できないこと
```

jujuko75 replied:
テスト内容追加しました。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

初回レビュー依頼にも関わらず非常に完成度が高いですね！
細かな部分にはなりますが、以下4点コメントいたしましたので、ご確認をお願いいたします。

jujuko75 committed:
コントローラー記載内容修正

メソッドの集約（before_action）

jujuko75 committed:
条件分岐記載内容修正（商品売却有無によるボタン表示）

jujuko75 committed:
単体テスト項目追加

・単体テスト項目（電話番号の英数混合）
・不要なテスト用ファイル削除

jujuko75 commented:
早速のレビューありがとうございます。
以下修正後の稼働確認のエビデンスです。ご確認よろしくお願いいたします。


**商品売却有無による条件分岐の記載方法修正**
【前提条件】
◆購入商品情報
・出品者：テスト次郎
・購入者：テスト次郎

【ログインユーザ＝テスト次郎】
①売却前：「編集／削除」ボタンが表示される、「購入画面に進む」ボタンは表示なし
　[https://gyazo.com/13bc523a2d8e07f9db608bfdb61a40e0](https://gyazo.com/13bc523a2d8e07f9db608bfdb61a40e0)
②売却後：「編集／削除／購入画面に進む」ボタン表示なし
　[https://gyazo.com/479d7ba0151a80b67568348d57b71bf2](https://gyazo.com/479d7ba0151a80b67568348d57b71bf2)


【ログインユーザ＝テスト太郎】
①売却前：「購入画面に進む」ボタンが表示され、「編集／削除」ボタンは表示なし
　[https://gyazo.com/23e63155e1b8a8951bbffb3c9fac1225](https://gyazo.com/23e63155e1b8a8951bbffb3c9fac1225)

②売却後：「編集／削除／購入画面に進む」ボタン表示なし
　[https://gyazo.com/929278b6e7fe42bf34a1af0151abf4e0](https://gyazo.com/929278b6e7fe42bf34a1af0151abf4e0)
　[https://gyazo.com/a4603be0e1ee9cce0bf1fd2fc644326d](https://gyazo.com/a4603be0e1ee9cce0bf1fd2fc644326d)
　

**単体テスト項目の追加**
・テスト項目（電話番号英数混合の場合保存できないこと）追加
[https://gyazo.com/6890f060a0d568642750452c757a92a9](https://gyazo.com/6890f060a0d568642750452c757a92a9)

kuu0123 commented on app/controllers/buyerinfos_controller.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  def redirect_to_root
+    set_item
```
```ruby
	    set_item
```
`set_item`はbefore_actionで呼び出しているため、こちらの記述は削除しましょう。
また、現状`set_item`よりも先に`redirect_to_root`が呼び出されているようなので、before_actionの呼び出し順を変更しましょう。

jujuko75 replied:
ご指摘の内容について修正しました。

kuu0123 commented on app/controllers/buyerinfos_controller.rb:
```diff
@@ -0,0 +1,48 @@
+class BuyerinfosController < ApplicationController
+  before_action :authenticate_user!, only:[:index, :create]
+  before_action :redirect_to_root, only:[:index]
```
```ruby
	  before_action :redirect_to_root, only:[:index]
```
こちらの記述はindex,createどちらにも適用させておきましょう。
検証ツールを使用した不正行為でcreateアクションにアクセスされるのを防ぎたいためです。

jujuko75 replied:
ご指摘の内容で修正しました。

kuu0123 commented on app/views/items/show.html.erb:
```diff
@@ -23,15 +23,17 @@
+        <%# if !@item.order.present?%>
     <%= link_to '商品の編集', edit_item_path(@item.id), method: :get, class: "item-red-btn" %>
     <p class='or-text'>or</p>
     <%= link_to '削除', item_path(@item.id), method: :delete, class:'item-destroy' %>
+        <%# end%>
       <% else %>
-    <%# 商品が売れていない場合はこちらを表示しましょう → 商品購入機能実装後 %>
-    <%= link_to '購入画面に進む', "#" ,class:"item-red-btn"%>
-    <%# //商品が売れていない場合はこちらを表示しましょう → 商品購入機能実装後 %>
+        <%# if !@item.order.present?%>
+        <%= link_to '購入画面に進む', item_order_buyerinfos_path(@item.id) ,class:"item-red-btn"%>
+        <%# end%>
```
上記計4行のコメントアウトは必要でしょうか？不要であれば削除しておきましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめします。

古いコードを記述したアプリを第三者に提出する（例：転職活動に使用される場合）のは好ましくありません。
また、捕手の観点からもチェックすべきコードが多くなるため、不要な記述は削除が好ましいです。

jujuko75 replied:
削除漏れ失礼しました。
稼働確認をしながら修正を行っていたもので、不要記述のため、削除しました。

kuu0123 commented on app/models/order_buyerinfo.rb:
```diff
@@ -0,0 +1,18 @@
+    validates :house_number
+    validates :phone_number,format: {with: /\A\d{10,11}\z/}
+    validates :token
+  end
```
「空では登録できない」のバリデーションに、`user_id`および`item_id`を追加しましょう！

こちらの記述がなくてもアプリ自体は正常に動作しますが、データベースに保存する前にバリデーションをかけることで、エラーを事前に確認できるようになります。
（Formオブジェクトのモデルでは、外部キーに`presence: true`が自動でかかりません）

またアプリ製作者がカラム名や保存するname属性を誤ってしまうという人為的なミスを減らすことができるなど、堅牢なアプリを作成するには必要な工程になります！

jujuko75 replied:
ご指摘ありがとうございます。
Formオブジェクトモデルとの違い点についてもご丁寧にありがとうございます。
変更が必要な理由も含めて確認し、修正しました。

kuu0123 commented on spec/models/order_buyerinfo_spec.rb:
```diff
@@ -0,0 +1,74 @@
+        @order_buyerinfo.valid?
+        expect(@order_buyerinfo.errors.full_messages).to include("Token can't be blank")
+      end
+    end
```
モデルにuser_idとitem_idのバリデーションを追加した後は、テストにも以下の項目を実装しましょう！
```
・user_idが空では購入できないこと
・item_idが空では購入できないこと
```

jujuko75 replied:
バリデーション追加に伴いテスト項目も追加しました。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加で5点コメントいたしましたので、ご確認お願いします！

jujuko75 committed:
指摘取り込み

・不要記述（コメントアウト）削除
・バリデーション（user_id,item_id）追加
・コントローラー記載内容修正（不要記述削除、メソッド適用アクション追加）

jujuko75 committed:
単体テスト追加（バリデーション追加）

jujuko75 commented:
ご指摘ありがとうございます。
再度お手数ですがレビューお願いいたします。

**コントローラー記述内容修正に伴う稼働確認**
【前提条件】
◆売却済商品情報
・出品者：テスト次郎
・購入ページ：http://localhost:3000/items/14/order/buyerinfos
　
【ログインユーザ＝テスト太郎】
・売却済商品の購入ページURLを直接入力した場合、トップページへ遷移する。
[https://gyazo.com/64b920e678bdae08e415fbeede5df61f](https://gyazo.com/64b920e678bdae08e415fbeede5df61f)



**単体テスト追加**
・テスト項目（user_id,item_idバリデーション）追加
[https://gyazo.com/d9baead3cc86cc3b92020814eb667755](https://gyazo.com/d9baead3cc86cc3b92020814eb667755)

Hide-Morishita commented on app/controllers/buyerinfos_controller.rb:
```diff
@@ -0,0 +1,47 @@
+  def create
+    @order_buyerinfo = OrderBuyerinfo.new(buyer_params)
+
+    if  
+      @order_buyerinfo.valid?
+      pay_item 
+      @order_buyerinfo.save
+      redirect_to root_path
+    else
+      render :index
+    end
```
細かい部分ではありますが、`@order_buyerinfo.valid?`をifの直後に記述しましょう。
```ruby
if  @order_buyerinfo.valid?
  pay_item 
  @order_buyerinfo.save
  redirect_to root_path
else
   render :index
end
```

下記は、if文の記述方法になります。
```ruby
if 条件式
  条件式がtrueの場合の処理
else
  条件式がfalseの場合の処理
end
```

jujuko75 replied:
修正しました。

Hide-Morishita commented on spec/factories/order_buyerinfos.rb:
```diff
@@ -0,0 +1,13 @@
+    building_name { '東京ハイツ' }
+    phone_number { '09012345678' }
+    token { 'tok_abcdefghijk00000000000000000' }
+    user_id { 1 }
+    item_id { 1 }
```
```ruby
user_id { 1 }
item_id { 1 }
```
order_buyerinfos.rbはフォームオブジェクトのファイルになるかと思います。
フォームオブジェクトは、テーブルがないためアソシエーションを組むことができません。そのため、外部キーという概念は成り立たないかと思います。

したがって、user_idとitem_idはorder_buyerinfos.rbで作成せず、削除し```spec/models/order_buyerinfo_spec.rb ```のbefore内でuserやitemのFactoryBotから情報を生成し、idを取得しましょう。
ご不明な点がございましたら、コネクトにてご質問いただければと存じます。

jujuko75 replied:
ご指摘いただいた内容で修正します。

Hide-Morishita reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

jujuko75 committed:
指摘取り込み

単体テスト（バリデーション追加）の修正

jujuko75 commented:
レビューありがとうございます。
ご指摘いただきました通り、単体テストを修正しました。
以下テスト結果です。よろしくお願いいたします。

**単体テスト結果**
https://gyazo.com/9af541c710e23b1e87f974e924ebbec2

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れさまでした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。

jujuko75 commented:
レビューありがとうございました。
マージして進めます。
